semgrep-core/src/core/Report.ml

```diff
@@ -1,9 +1,79 @@
-(* Full result information *)
-(* TODO: We could use atdgen to specify those to factorize type definitions
+(********************************************************************************
+ * Full result information
+ *
+ * In addition to the results (matches + errors), we also may report extra
+ * information, such as the skipped targets or the profiling times. Many
+ * of the types here are created to collect profiling information in as
+ * well-typed a manner as possible. Creating a type for practically every
+ * stage that reports matches is annoying, but prevents us from relying on
+ * dummy values or unlabeled tuples.
+ *
+ * Another challenge we face is that the extra information can take a lot
+ * of memory, which scales with both the number of rules and files. On
+ * large repos, this is the most significant factor driving semgrep's
+ * memory consumption. Therefore, if the skipped targets (for example) are
+ * not being used, we don't want to save them. On the other hand, we want
+ * to feel confident in the correctness of the code and make it easy to
+ * know what is or isn't being saved. And we want our code to be relatively
+ * readable.
+ *
+ * The debug_info type attempts to solve this. It has a variant for each
+ * verbosity mode semgrep-core may be invoked with, which contains all
+ * the fields semgrep requests from semgrep-core in that mode. Each result
+ * contains debug_info in addition to the always-reported fields. The
+ * variant of debug_info used in the results is always determined either
+ * by the mode, which is a global set in Main.ml after the arguments are
+ * read, or by a previous result. In this way we ensure that the fields
+ * stored in the result are determined by the arguments passed by the user.
+ *
+ * Alternatives considered to the extra field:
+ * - storing the information but just ommitting it in the final result (I
+ *   tried this but it still uses too much memory)
+ * - using option types within the result types for field and a global
+ *   config to decide which fields are in use
+ *
+ * I considered the latter, but I'm not a fan of mix-and-match option types.
```

Comment:
You can also use `option` types without giving the user the choice of picking exactly what info they want.

Reply:
I like the approach you took, although the part of "giving the user the choice" kind of confused me, because I think it's an orthogonal issue.

Reply:
The reason I said that is because if we were to give the user choice, then we would need to make every combination of options possible. In that case, I think it would make more sense to use option types.